### PR TITLE
Typescript migration

### PR DESCRIPTION
**Pull request by aime-risson**
Migrate libjs to typescript.

All the unit tests passed. The lib was tested on node, browser and react.
Also added a react example.

The bundle for browser is separated from the rest as I couldn't figure a way to bundle everything into one file and still be compatible with node, browser and react

### CODE REVIEW

File: src/transaction_builder.ts [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    maybeHexToUint8Array,
+    maybeStringToUint8Array,
+    toByteArray,
+    uint8ArrayToHex
+} from "./utils.js";
+import {deriveAddress, deriveKeyPair, sign} from "./crypto.js";
+
+const VERSION = 1
+
+const txTypes = {
```
Comment (samuelmanzanera) [inline src/transaction_builder.ts:22]: Is this variable still relevant as you used an enum for the tx type ?

File: src/types.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    transfer = 'transfer',
+    hosting = 'hosting',
+    token = 'token',
+    data = 'data',
+    contract = 'contract',
+    code_proposal = 'code_proposal',
+    code_approval = 'code_approval',
+}
+
+enum NetworkTypeTransaction {
```
Comment (samuelmanzanera) [inline src/types.ts:27]: Why do you need this kind of enum ?

File: src/api.ts [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+import {maybeUint8ArrayToHex} from "./utils.js";
+import {Balance, NearestEndpoint, OracleData, Ownership, Token, TransactionFee} from "./types.js";
+import TransactionBuilder from "./transaction_builder.js";
+
+/**
+ * Send a custom query to the Archethic API
+ * @param query
+ * @param endpoint
+ */
+export async function rawGraphQLQuery(query: string, endpoint: string) {
```
Comment (samuelmanzanera) [inline src/api.ts:12]: maybe you should also return type a Promise

File: src/types.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    amount: number,
+    from: string,
+    spent: boolean,
+    timestamp: number,
+    tokenAddress: string,
+    tokenId: number,
+    type: string,
+}
+
+type ValidationStamp = {
```
Comment (samuelmanzanera) [inline src/types.ts:137]: Same here ?

File: dist/types.d.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    name: string;
+    type: string;
+    symbol: string;
+    supply: number;
+    decimals: number;
+    properties?: Object[];
+    collection?: Object[];
+    ownerships?: Ownership[];
+};
+type CrossValidationStamp = {
```
Comment (samuelmanzanera) [inline dist/types.d.ts:52]: Same here?

File: src/types.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    type: string,
+    symbol: string,
+    supply: number,
+    decimals: number,
+    properties?: Object[],
+    collection?: Object[],
+    ownerships?: Ownership[],
+}
+
+type CrossValidationStamp = {
```
Comment (samuelmanzanera) [inline src/types.ts:74]: Same here?

File: src/transaction_builder.ts (hunk 1 above)
Comment (aime-risson) [reply src/transaction_builder.ts:22]: I think yes because I need the tx type number and not the it's name

File: src/types.ts (hunk 2 above)
Comment (aime-risson) [reply src/types.ts:27]: libjs doesn't require it but developers might need it

File: src/types.ts (hunk 4 above)
Comment (aime-risson) [reply src/types.ts:137]: to cast queried transaction

File: src/transaction_builder.ts (hunk 1 above)
Comment (samuelmanzanera) [reply src/transaction_builder.ts:22]: ok so you can have multiple alternatives to still use type system:
- use another enum 
- replace enum by an object
- create a function which take an enum member and return its number

Comment (aime-risson) [reply src/transaction_builder.ts:22]: I think best option is the function

**Review by bchamagne — CHANGES_REQUESTED** [review]
Sorry for reviewing this so late.
I could not test everything yet but here's a few thing I noticed:

# examples
with current code, only `browserConnect` works. I had to comment these 3 lines in the package.json to make all example work: 

```
-    "browser": "./dist/archethic.browser.js",
-    "unpkg":  "dist/archethic.browser.js",
-    "jsdelivr": "./dist/archethic.browser.js",
```

If I understand correctly these are not needed anyway because a browser app that does not use a bundler will not even look at the package.json and will just `<script src=archethic-browser.js></script>`

What do you think?

# react example 

```
Module not found: Error: Can't resolve 'crypto' in '/home/bastien/Chamagne/archethic/libjs/node_modules/sjcl'
BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "crypto": require.resolve("crypto-browserify") }'
	- install 'crypto-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "crypto": false }
```
I get this error, but the app is still displayed. I have no idea if the app works because I am unable to install the archethic-wallet locally :sob: .

ps: I am using node v19.2

**Comment by samuelmanzanera** [comment]
@bchamagne This is working for me as well, but I didn't check with the wallet either.

**Review by samuelmanzanera — APPROVED** [review]

**Review by bchamagne — APPROVED** [review]

**Pull request merged**